microsoft/testsuites/dpdk/common.py

```diff
@@ -17,6 +22,261 @@
+                break
+        assert_that(str(self._source_path)).described_as(
+            "Source path was not set by tar installer, "
+            "likely an unexpected file extension for "
+            f"the tar file: {self.tar_filename}"
+        ).is_not_equal_to(str(work_path))
+        # unpack into the dest dir
+        node.tools[Tar].extract(
+            file=str(remote_path),
+            dest_dir=str(work_path),
+            gzip=True,
+        )
+        return self._source_path
+
+
+class Installer:
+    # Generic 'Installer' parent class for DpdkTestpmd/rdma-core
+    # NOTE: This should not be instantiated directly.
+    _err_msg = "not implemented for this installation type."
+
+    # setup the node before starting
+    # ex: updating the kernel, enabling features, checking drivers, etc.
+    def _setup_node(self) -> None:
+        raise NotImplementedError(f"_setup_node {self._err_msg}")
+
+    # check if the package is already installed:
+    # Is the package installed from source? Or from the package manager?
+    # Does the version match the one we want if we need a specific one?
+    def _check_if_installed(self) -> bool:
+        raise NotImplementedError(f"_check_if_installed {self._err_msg}")
+
+    # setup the installation (install Ninja, Meson, etc)
+    def _setup_installation(self) -> None:
+        if self._downloader:
+            self._asset_path = self._downloader.download()
+        else:
+            self._node.log.debug("No downloader assigned to installer.")
+
+    # do the build and installation
+    def _run_installation(self) -> None:
+        raise NotImplementedError(f"_run_installation {self._err_msg}")
+
+    # remove an installation
+    def _clean_previous_installation(self) -> None:
+        raise NotImplementedError(f"_clean_previous_installation {self._err_msg}")
+
+    # provide an opportunity to check tags, fetch a subproject,
+    # modify the config, etc.
+    def _configure_installation(self) -> None:
+        raise NotImplementedError(f"_configure_installation {self._err_msg}")
+
+    # install the dependencies
+    def _install_dependencies(self) -> None:
+        if self._os_dependencies is not None:
+            self._os_dependencies.install_required_packages(
+                self._node, extra_args=self._package_manager_extra_args
+            )
+
+    # define how to check the installed version
+    def get_installed_version(self) -> VersionInfo:
+        raise NotImplementedError(f"get_installed_version {self._err_msg}")
+
+    # run the defined setup and installation steps.
+    def do_installation(self, required_version: Optional[VersionInfo] = None) -> None:
+        self._setup_node()
+        if self._check_if_installed() and required_version is not None:
+            if self.get_installed_version() == required_version:
+                return
+            else:
+                self._clean_previous_installation()
+        self._install_dependencies()
```

Comment:
Don't split it to multiple methods, try to overwrite methods of base classes.

Reply:
I'm fine with fixing it later to suit your preference but I'm not doing that for this commit. The Installer is not used outside of DpdkTestpmd, so it's fine to merge as-is now. I'm going to touch this again when I implement the rdma-core Installer, so I'll change it in that PR.